File: apps/android/feature/tasks/src/main/kotlin/dev/optio/feature/tasks/data/RunModels.kt
package dev.optio.feature.tasks.data

import dev.optio.core.ui.format.Cost
import java.time.Instant
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

// Row shapes returned by the task-related routes (iOS `Features/Run/Tasks/RunModels.swift`).
// Routes enrich rows beyond the generated `OptioTask` (prNumber, prChecksStatus, costUsd,
// isStalled, …) and several fields are nullable in practice, so these are declared locally with
// everything optional; see apps/api/src/schemas/task.ts. Dates never fail a decode
// ([LenientInstantSerializer]).

/** A task row (`tasks`), as the detail, list, subtask, dependency and blueprint-run routes return it. */
@Serializable
data class TaskRow(
    val id: String,
    val title: String = "",
    val prompt: String? = null,
    val repoUrl: String? = null,
    val repoBranch: String? = null,
    val state: String = "",
    val agentType: String? = null,
    val sessionId: String? = null,
    val prUrl: String? = null,
    val prNumber: Int? = null,
    val prState: String? = null,
    val prChecksStatus: String? = null,
    val prReviewStatus: String? = null,
    val resultSummary: String? = null,
    val costUsd: String? = null,
    val inputTokens: Long? = null,
    val outputTokens: Long? = null,
    val modelUsed: String? = null,
    val errorMessage: String? = null,
    val ticketSource: String? = null,
    val ticketExternalId: String? = null,
    val retryCount: Int? = null,
    val maxRetries: Int? = null,
    val priority: Int? = null,
    val parentTaskId: String? = null,
    val taskType: String? = null,
    val blocksParent: Boolean? = null,
    val isStalled: Boolean? = null,
    val pendingReason: String? = null,
    /** `cluster` (an Optio pod) or `local` (a terminal on the owner's machine). */
    val runTarget: String? = null,
    val localDir: String? = null,
    /** Local runs: the `local_terminals` row executing this task. */
    val localTerminalId: String? = null,
    @Serializable(with = LenientInstantSerializer::class) val lastActivityAt: Instant? = null,
    val activitySubstate: String? = null,
    @Serializable(with = LenientInstantSerializer::class) val createdAt: Instant? = null,
    @Serializable(with = LenientInstantSerializer::class) val updatedAt: Instant? = null,
    @Serializable(with = LenientInstantSerializer::class) val startedAt: Instant? = null,
    @Serializable(with = LenientInstantSerializer::class) val completedAt: Instant? = null,
) {
    /** `owner/repo` (iOS `repoShortName`). */
    val repoShortName: String
        get() = RunFormatting.repoShortName(repoUrl.orEmpty())

    /** "PR #519" once a PR exists. */
    val prLabel: String?
        get() {
            if (prUrl == null) return null
            if (prNumber != null) return "PR #$prNumber"
            val last = prUrl.trimEnd('/').substringAfterLast('/', "")
            return if (last.isNotEmpty()) "PR #$last" else "PR"
        }

    /** "#519" for rows. */
    val prNumberLabel: String?
        get() = prNumber?.let { "#$it" } ?: prUrl?.trimEnd('/')?.substringAfterLast('/', "")?.takeIf { it.isNotEmpty() }?.let { "#$it" }

    val costText: String?
        get() = Cost.formatIfNonZero(costUsd)

    /** Runs on the owner's machine (Optio Local) rather than in a pod. */
    val isLocal: Boolean
        get() = runTarget == "local"

    /** Running with no recent activity (the list routes' `isStalled`). */
    val isStalledRunning: Boolean
        get() = isStalled == true && state == "running"
}

/** `GET /api/tasks/:id`: the task plus its enrichment. */
@Serializable
data class TaskDetailResponse(
    val task: TaskRow,
    val pendingReason: String? = null,
    val stallInfo: StallInfoRow? = null,
)

/** The silent-activity detector's output for a running task. */
@Serializable
data class StallInfoRow(
    val isStalled: Boolean = false,
    val silentForMs: Double = 0.0,
    val thresholdMs: Double? = null,
    val lastLogSummary: String? = null,
)

/** A `task_events` row (`GET /api/tasks/:id/events`); only the trigger drives UI (plan review). */
@Serializable
data class TaskEventRow(
    val id: String = "",
    val fromState: String? = null,
    val toState: String? = null,
    val trigger: String? = null,
    val message: String? = null,
    @Serializable(with = LenientInstantSerializer::class) val createdAt: Instant? = null,
)

/** Item of `GET /api/tasks/:id/activity`: comments, state events and messages merged. */
@Serializable
data class TaskActivityItem(
    val type: String = "",
    val id: String,
    val taskId: String? = null,
    @Serializable(with = LenientInstantSerializer::class) val createdAt: Instant? = null,
    val content: String? = null,
    val user: ActivityUser? = null,
    val fromState: String? = null,
    val toState: String? = null,
    val trigger: String? = null,
    val message: String? = null,
    val userId: String? = null,
    val mode: String? = null,
    @Serializable(with = LenientInstantSerializer::class) val deliveredAt: Instant? = null,
    @Serializable(with = LenientInstantSerializer::class) val ackedAt: Instant? = null,
) {
    @Serializable
    data class ActivityUser(
        val id: String? = null,
        val displayName: String? = null,
        val avatarUrl: String? = null,
    )

    val isComment: Boolean
        get() = type == "comment"
}

/** A repo from `GET /api/repos` (the scheduled form's picker). */
@Serializable
data class RunRepoRow(
    val id: String,
    val repoUrl: String,
    val fullName: String? = null,
    val defaultBranch: String? = null,
    val defaultAgentType: String? = null,
) {
    val displayName: String
        get() = fullName ?: RunFormatting.repoShortName(repoUrl)
}

/** A prompt template from `GET /api/prompt-templates?kind=…`. */
@Serializable
data class RunPromptTemplateRow(
    val id: String,
    val name: String,
    val template: String = "",
    val kind: String? = null,
    val description: String? = null,
    val defaultAgentType: String? = null,
)

/** A scheduled Task blueprint (`task_configs` row). */
@Serializable
data class TaskConfigRow(
    val id: String,
    val name: String = "",
    val description: String? = null,
    val title: String = "",
    val prompt: String = "",
    val promptTemplateId: String? = null,
    val repoUrl: String = "",
    val repoBranch: String? = null,
    val agentType: String? = null,
    val maxRetries: Int? = null,
    val priority: Int? = null,
    val enabled: Boolean = true,
    val runTarget: String? = null,
    val localDir: String? = null,
    val agentOptions: Map<String, JsonElement>? = null,
    @Serializable(with = LenientInstantSerializer::class) val createdAt: Instant? = null,
    @Serializable(with = LenientInstantSerializer::class) val updatedAt: Instant? = null,
)

/** Shared formatting of the Run screens (iOS `RunFormatting`). */
object RunFormatting {
    /** `https://github.com/acme/web.git` → `acme/web` (drops the scheme and the host). */
    fun repoShortName(url: String): String {
        var s = url
        val scheme = s.indexOf("://")
        if (scheme >= 0) s = s.substring(scheme + 3)
        val slash = s.indexOf('/')
        if (slash >= 0) s = s.substring(slash + 1)
        if (s.endsWith(".git")) s = s.dropLast(4)
        return s
    }

    /** Agent runtimes, value → label (the web's agent picker order). */
    val agentTypes: List<Pair<String, String>> = listOf(
        "claude-code" to "Claude Code",
        "codex" to "OpenAI Codex",
        "copilot" to "GitHub Copilot",
        "opencode" to "OpenCode",
        "gemini" to "Google Gemini",
        "openclaw" to "OpenClaw",
        "cursor" to "Cursor",
    )

    /** "Claude Code"; null → "default agent" (a blueprint that uses the repo's default). */
    fun agentLabel(type: String?): String {
        if (type == null) return "default agent"
        return agentTypes.firstOrNull { it.first == type }?.second ?: type
    }

    /** "45s", "3m 12s", "1h 4m" from milliseconds. */
    fun duration(ms: Double): String {
        val s = (ms / 1000).toLong().coerceAtLeast(0)
        if (s >= 3600) return "${s / 3600}h ${(s % 3600) / 60}m"
        if (s >= 60) return "${s / 60}m ${s % 60}s"
        return "${s}s"
    }

    /** The web task list's stage (pipeline-timeline.tsx): queue, setup, running, ci, review, … */
    fun stage(task: TaskRow): String = when (task.state) {
        "completed", "cancelled" -> "done"
        "failed" -> "failed"
        "pending", "queued", "waiting_on_deps" -> "queue"
        "provisioning" -> "setup"
        "running" -> "running"
        "needs_attention" -> "attention"
        "pr_opened" -> when {
            task.prReviewStatus != null && task.prReviewStatus !in setOf("none", "pending") -> "review"
            task.prChecksStatus == "passing" -> "review"
            else -> "ci"
        }
        else -> "queue"
    }
}

File: apps/ios/Optio/Features/Run/Tasks/RunModels.swift
import Foundation

// Row shapes returned by the task-related routes. Routes enrich rows beyond the
// generated `OptioTask` (prNumber, prChecksStatus, costUsd, isStalled, …), and
// several fields are nullable in practice, so these are declared locally with
// everything optional. See apps/api/src/schemas/task.ts.

struct TaskRow: Decodable, Identifiable, Hashable {
    let id: String
    var title: String
    var prompt: String?
    var repoUrl: String?
    var repoBranch: String?
    var state: String
    var agentType: String?
    var sessionId: String?
    var prUrl: String?
    var prNumber: Int?
    var prState: String?
    var prChecksStatus: String?
    var prReviewStatus: String?
    var resultSummary: String?
    var costUsd: String?
    var inputTokens: Int?
    var outputTokens: Int?
    var modelUsed: String?
    var errorMessage: String?
    var ticketSource: String?
    var ticketExternalId: String?
    var retryCount: Int?
    var maxRetries: Int?
    var priority: Int?
    var parentTaskId: String?
    var taskType: String?
    var blocksParent: Bool?
    var isStalled: Bool?
    var pendingReason: String?
    var lastActivityAt: Date?
    var activitySubstate: String?
    var createdAt: Date?
    var updatedAt: Date?
    var startedAt: Date?
    var completedAt: Date?

    var taskState: TaskState { TaskState(rawValue: state) ?? .unknown }
    var repoShortName: String { RunFormatting.repoShortName(repoUrl ?? "") }
    var prLabel: String? {
        guard prUrl != nil else { return nil }
        if let prNumber { return "PR #\(prNumber)" }
        if let n = prUrl?.split(separator: "/").last { return "PR #\(n)" }
        return "PR"
    }
    var costText: String? {
        guard let costUsd, let v = Double(costUsd) else { return nil }
        return Cost.format(v)
    }
}

struct RunTaskStats: Decodable {
    var total: Int = 0
    var queued: Int = 0
    var running: Int = 0
    var ci: Int = 0
    var review: Int = 0
    var needsAttention: Int = 0
    var failed: Int = 0
    var completed: Int = 0
}

/// Row from `GET /api/tasks/:id/logs` (schemas/task.ts LogEntrySchema).
struct TaskLogRow: Decodable {
    var id: String?
    var stream: String?
    var content: String
    var logType: String?
    var metadata: [String: AnyCodable]?
    var timestamp: String?

    func asEntry(taskId: String) -> AgentLogEntry {
        AgentLogEntry(
            taskId: taskId,
            timestamp: timestamp ?? "",
            type: AgentLogEntry.TypeValue(rawValue: logType ?? "text") ?? .text,
            content: content,
            metadata: metadata
        )
    }
}

/// Frame from `/ws/logs/:taskId` (ws/log-stream.ts). The generated `TaskLogEvent`
/// omits `logType`/`metadata`/`catchUp`, which the log renderer needs.
struct TaskLogFrame: Decodable {
    var type: String
    var taskId: String?
    var content: String?
    var stream: String?
    var timestamp: String?
    var logType: String?
    var metadata: [String: AnyCodable]?
    var catchUp: Bool?
}

/// Item from `GET /api/tasks/:id/activity` — comments, events and messages merged.
struct TaskActivityItem: Decodable, Identifiable {
    struct User: Decodable { var id: String?; var displayName: String?; var avatarUrl: String? }
    var type: String
    var id: String
    var taskId: String?
    var createdAt: Date?
    var content: String?
    var user: User?
    var fromState: String?
    var toState: String?
    var trigger: String?
    var message: String?
    var userId: String?
    var mode: String?
    var deliveredAt: Date?
    var ackedAt: Date?
}

struct RunRepoRow: Decodable, Identifiable, Hashable {
    let id: String
    var repoUrl: String
    var fullName: String?
    var defaultBranch: String?
    var defaultAgentType: String?
    var displayName: String { fullName ?? RunFormatting.repoShortName(repoUrl) }
}

struct RunPromptTemplateRow: Decodable, Identifiable, Hashable {
    let id: String
    var name: String
    var template: String
    var kind: String?
    var description: String?
    var defaultAgentType: String?
}

/// Trigger row shared by task-configs and the unified /api/tasks/:id/triggers routes.
struct TriggerRow: Decodable, Identifiable, Hashable {
    let id: String
    var targetType: String?
    var targetId: String?
    var type: String
    var config: [String: AnyCodable]?
    var paramMapping: [String: AnyCodable]?
    var enabled: Bool
    var lastFiredAt: Date?
    var nextFireAt: Date?
    var createdAt: Date?

    var cronExpression: String? { config?["cronExpression"]?.stringValue }
    var webhookPath: String? { config?["path"]?.stringValue }
    var ticketSource: String? { config?["source"]?.stringValue }
    var ticketLabels: [String] { config?["labels"]?.arrayValue?.compactMap { $0.stringValue } ?? [] }

    var summary: String {
        switch type {
        case "schedule": return cronExpression ?? "cron"
        case "webhook": return webhookPath.map { "/api/hooks/\($0)" } ?? "webhook"
        case "ticket":
            let src = ticketSource ?? "github"
            return ticketLabels.isEmpty ? src : "\(src) · \(ticketLabels.joined(separator: ", "))"
        default: return "manual"
        }
    }
}

struct TaskConfigRow: Decodable, Identifiable, Hashable {
    let id: String
    var name: String
    var description: String?
    var title: String
    var prompt: String
    var promptTemplateId: String?
    var repoUrl: String
    var repoBranch: String?
    var agentType: String?
    var maxRetries: Int?
    var priority: Int?
    var enabled: Bool
    var createdAt: Date?
    var updatedAt: Date?
}

/// Row from `GET /api/issues`. Shape varies by provider (schemas/session.ts).
struct IssueRow: Decodable, Identifiable {
    struct Repo: Decodable { var id: String?; var fullName: String?; var repoUrl: String? }
    struct OptioTaskRef: Decodable { var taskId: String?; var state: String? }
    var id: AnyCodable?
    var number: AnyCodable?
    var title: String
    var body: String?
    var state: String?
    var url: String?
    var labels: [String]?
    var author: String?
    var assignee: String?
    var source: String?
    var hasOptioLabel: Bool?
    var repo: Repo?
    var optioTask: OptioTaskRef?
    var createdAt: String?
    var updatedAt: String?

    var numberText: String {
        if let n = number?.intValue { return "#\(n)" }
        if let d = number?.doubleValue { return "#\(Int(d))" }
        if let s = number?.stringValue { return s }
        return ""
    }
    var numberInt: Int? { number?.intValue ?? number?.doubleValue.map { Int($0) } }
    var isAssignable: Bool {
        (source == nil || source == "github" || source == "gitlab") && repo?.id != nil && optioTask == nil
    }
    var identity: String { "\(repo?.fullName ?? source ?? "")-\(numberText)-\(id?.stringValue ?? id?.doubleValue.map { String($0) } ?? "")" }
}

enum RunFormatting {
    static func repoShortName(_ url: String) -> String {
        var s = url
        if let r = s.range(of: "://") { s = String(s[r.upperBound...]) }
        if let slash = s.firstIndex(of: "/") { s = String(s[s.index(after: slash)...]) }
        if s.hasSuffix(".git") { s.removeLast(4) }
        return s
    }

    static let agentTypes: [(String, String)] = [
        ("claude-code", "Claude Code"),
        ("codex", "OpenAI Codex"),
        ("copilot", "GitHub Copilot"),
        ("opencode", "OpenCode"),
        ("gemini", "Google Gemini"),
        ("openclaw", "OpenClaw"),
        ("cursor", "Cursor"),
    ]

    static func agentLabel(_ type: String?) -> String {
        guard let type else { return "default agent" }
        return agentTypes.first { $0.0 == type }?.1 ?? type
    }

    static func duration(_ ms: Double) -> String {
        let s = Int(ms / 1000)
        if s >= 3600 { return "\(s / 3600)h \((s % 3600) / 60)m" }
        if s >= 60 { return "\(s / 60)m \(s % 60)s" }
        return "\(s)s"
    }

    /// Mirrors the web's task-list stage derivation (pipeline-timeline.tsx).
    static func stage(for t: TaskRow) -> String {
        switch t.state {
        case "completed", "cancelled": return "done"
        case "failed": return "failed"
        case "pending", "queued", "waiting_on_deps": return "queue"
        case "provisioning": return "setup"
        case "running": return "running"
        case "needs_attention": return "attention"
        case "pr_opened":
            if let r = t.prReviewStatus, !["none", "pending"].contains(r) { return "review" }
            if t.prChecksStatus == "passing" { return "review" }
            return "ci"
        default: return "queue"
        }
    }
}
